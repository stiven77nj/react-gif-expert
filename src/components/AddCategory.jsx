import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ nuevaCategoria }) => {

    // Hook useState.
    // Manejar el estado del input
    const [valorInput, setValorInput] = useState('');


    // Funcion para manejar los cambios en el input.
    // Recibe el 'event' del input.
    // Se obtiene el 'value' para obtener los cambios.
    const cambioDelInput = ( event ) => {
        const { value } = event.target;
        setValorInput( value );
    }


    // Funcion para el posteo del formulario
    const posteoFormulario = ( event ) => {
        event.preventDefault(); // No se recarga el navegador

        // Se valida un numero minimo de caracteres en el input para que sea agregada
        // la nueva categoria
        if ( valorInput.trim().length <= 3 ) {
            return;
        }

        // Se envia el nuevo valor a la funcion de agregarCategoria,
        // para que sea agregado a la lista de categorias.
        nuevaCategoria( valorInput );
        setValorInput(''); // Reseteo del formulario
    }


    return (
        <>

            <form
                onSubmit={ posteoFormulario }
                aria-label='form'
            >
                <input
                    type="text"
                    placeholder='Buscar gifs'
                    value={ valorInput }
                    onChange={ cambioDelInput }
                />
            </form>

        </>
    )
}

// PropTypes
AddCategory.propTypes = {
    nuevaCategoria: PropTypes.func.isRequired
}