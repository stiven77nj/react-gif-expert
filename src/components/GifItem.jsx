import React from 'react';
import PropTypes from 'prop-types';

export const GifItem = ({ title, url }) => {
    return (
        <>
            <div className='card'>
                <h5>{title}</h5>
                <img src={url} alt={title} />
            </div>
        </>
    )
}


// PropTypes
GifItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}