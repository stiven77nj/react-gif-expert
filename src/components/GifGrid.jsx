import React from 'react';

import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({ categoria }) => {

    // Se llama al custom hook para obtener la data y dibujarla en pantalla
    const { data, isLoading } = useFetchGifs( categoria ); 

    return (
        <>
            {/* Titulo de la categoria */}
            <h3>{categoria}</h3>
            <div className='card-grid' style={{ background: '#ccc'}}>
                {
                    !isLoading /* Si isLoading es false */
                    
                    ?

                    data.map(d => (
                        <GifItem
                            key={d.id}
                            {...d} /* Enviamos todas las propiedades con ayuda del operador ... */
                        />
                    ))

                    :

                    <h3>Cargando...</h3> /* Si isLoading es true */
                }
            </div>
        </>
    )
}

// Cada vez que react detecta un cambio vuelve a redibujar el componente.
// Hay acciones que se le pueden indicar a react para evitar este comportamiento.
// useEffect sirve para disparar efectos secundarios cuando algo ocurra