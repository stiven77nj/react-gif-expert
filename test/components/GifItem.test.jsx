import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe('Pruebas de <GifItem />', () => {

    const title = 'Saitama';
    const url = 'https://one-punch.com/saitama.jpg';
    
    test('Debe hacer match con el snapshot', () => {
        // El componente debe de recibir las propidades de "title" y "url"
        const { container } = render( <GifItem title={title} url={url} /> ); // Renderizar el componente
        expect( container ).toMatchSnapshot(); // Debe hacer match
    });

    test('Debe de mostrar la imagen con el url y el alt indicado', () => {
        render( <GifItem title={title} url={url} /> ); // Renderizar el componente
        const { src, alt } = screen.getByRole('img'); // Se obtiene el 'url' y 'alt' del elemento 'img' del componente
        expect( src ).toBe( url ); // Comparando las 'url'
        expect( alt ).toBe( title ); // Comparando ls 'alt'
    });

    test('Debe de mostrar el titulo en el componente', () => {
        render( <GifItem title={title} url={url} /> ); // Renderizar el componente
        expect( screen.getByText( title ) ).toBeTruthy(); // Se evalua si el 'title' esta como un texto en el componente
    });
    
});