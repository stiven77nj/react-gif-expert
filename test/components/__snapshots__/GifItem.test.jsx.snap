// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <GifItem /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <h5>
      Saitama
    </h5>
    <img
      alt="Saitama"
      src="https://one-punch.com/saitama.jpg"
    />
  </div>
</div>
`;
